File: iOS/TUILiveKit/Sources/LiveIdentityGenerator.swift
//
//  LiveIdentityGenerator.swift
//  TUILiveKit
//
//  Created by WesleyLei on 2024/6/6.
//

import Foundation

@objcMembers
public final class LiveIdentityGenerator: NSObject {
    @objc public enum RoomType: Int, CaseIterable {
        case live
        case voice
        case ktv
        case unknown
    }

    public static let shared = LiveIdentityGenerator()
    private override init() {
    }

    public func generateId(_ id: String, type: RoomType) -> String {
        return getPrefix(type: type) + id
    }

    public func getIDType(_ id: String) -> RoomType {
        for roomType in RoomType.allCases {
            if id.hasPrefix(getPrefix(type: roomType)) {
                return roomType
            }
        }
        return .unknown
    }
    
    private func getPrefix(type: RoomType) -> String {
        switch type {
        case .live:
            return "live_"
        case .voice:
            return "voice_"
        case .ktv:
            return "ktv_"
        case .unknown:
            return ""
        }
    }
}

File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/common/LiveIdentityGenerator.java
package com.trtc.uikit.livekit.common;

import androidx.annotation.NonNull;

public final class LiveIdentityGenerator {
    private static final LiveIdentityGenerator INSTANCE = new LiveIdentityGenerator();

    public static LiveIdentityGenerator getInstance() {
        return INSTANCE;
    }

    private LiveIdentityGenerator() {
    }

    public String generateId(String id, @NonNull RoomType type) {
        return type.getPrefix() + id;
    }

    public RoomType getIDType(String id) {
        if (id == null) {
            return null;
        }
        for (RoomType type : RoomType.values()) {
            if (id.startsWith(type.getPrefix())) {
                return type;
            }
        }
        return null;
    }

    public enum RoomType {
        LIVE("live_"),
        VOICE("voice_"),
        KTV("ktv_");

        private final String prefix;

        RoomType(String prefix) {
            this.prefix = prefix;
        }

        public String getPrefix() {
            return prefix;
        }
    }
}
